src/storage/redis.rs

```diff
@@ -0,0 +1,400 @@
+            conn.ltrim(&queue, self.max_commands, -1).await?;
+        }
+
+        Ok(())
+    }
+
+    async fn forward(&self, msg: &Message) -> Result<()> {
+        let mut conn = self.get_connection().await?;
+
+        // add to backlog
+        let key = self.prefixed(Queue::Backlog(&msg.id));
+        conn.set_ex(&key, msg, self.ttl as usize).await?;
+
+        // push to forward for every destination
+        let queue = self.prefixed(Queue::Forward);
+        for destination in &msg.destination {
+            let forwarded = ForwardedMessage {
+                id: msg.id.to_owned(),
+                destination: *destination,
+            };
+            conn.rpush(&queue, &forwarded).await?
+        }
+
+        Ok(())
+    }
+
+    async fn reply(&self, msg: &Message) -> Result<()> {
+        let mut conn = self.get_connection().await?;
+
+        conn.rpush(&msg.reply, msg).await?;
+        Ok(())
+    }
+
+    async fn local(&self) -> Result<Message> {
+        let mut conn = self.get_connection().await?;
+        let queue = self.prefixed(Queue::Local);
+        let ret: (Vec<u8>, Vec<u8>) = conn.blpop(&queue, 0).await?;
+
+        let (_, value) = ret;
+        let msg = Message::from_json(&value)?;
+        Ok(msg)
+    }
+
+    async fn queued(&self) -> Result<QueuedMessage> {
+        let mut conn = self.get_connection().await?;
+        let forward_queue = self.prefixed(Queue::Forward);
+        let reply_queue = self.prefixed(Queue::Reply);
+        let queues = (forward_queue, reply_queue);
+
+        let result;
+
+        loop {
+            let ret: (String, Vec<u8>) = conn.blpop(&queues, 0).await?;
+            let (queue, value) = ret;
+
+            match queue {
+                forward_queue => {
+                    let forwarded = ForwardedMessage::from_bytes_pair(&value)?;
+
+                    let mut ret = self.get_cached(&forwarded.id).await;
+                    let not_cached = ret.is_none();
+
+                    if not_cached {
+                        ret = self.get(&forwarded.id).await?;
+                    }
+
+                    match ret {
+                        Some(mut msg) => {
+                            msg.destination = vec![forwarded.destination];
+                            if not_cached {
+                                self.cache(&msg.id, &msg);
+                            }
+                            result = QueuedMessage::Forward(msg);
+                            break;
+                        }
+                        None => {
+                            log::debug!("message of {} has been expired", forwarded.id);
+                        }
+                    }
+                }
+
+                reply_queue => {
+                    // reply queue had the message itself
+                    // decode it directly
+                    let msg = Message::from_json(&value)?;
+                    result = QueuedMessage::Reply(msg);
+                    break;
```

Comment:
no need to assign then break, just return
```suggestion
                    return Ok(QueuedMessage::Reply(msg));
```